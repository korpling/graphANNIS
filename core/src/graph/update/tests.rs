use insta::assert_snapshot;

use super::*;

#[test]
fn serialize_deserialize_bincode() {
    let example_updates = vec![
        UpdateEvent::AddNode {
            node_name: "parent".into(),
            node_type: "corpus".into(),
        },
        UpdateEvent::AddNode {
            node_name: "child".into(),
            node_type: "corpus".into(),
        },
        UpdateEvent::AddEdge {
            source_node: "child".into(),
            target_node: "parent".into(),
            layer: "annis".into(),
            component_type: "PartOf".into(),
            component_name: "".into(),
        },
    ];

    let mut updates = GraphUpdate::new();
    for e in example_updates.iter() {
        updates.add_event(e.clone()).unwrap();
    }

    let seralized_bytes: Vec<u8> = bincode::serialize(&updates).unwrap();
    let deseralized_update: GraphUpdate = bincode::deserialize(&seralized_bytes).unwrap();

    assert_eq!(3, deseralized_update.len().unwrap());
    let deseralized_events: Vec<UpdateEvent> = deseralized_update
        .iter()
        .unwrap()
        .map(|e| e.unwrap().1)
        .collect();
    assert_eq!(example_updates, deseralized_events);
}

#[test]
fn serialize_deserialize_bincode_empty() {
    let example_updates: Vec<UpdateEvent> = Vec::new();

    let mut updates = GraphUpdate::new();
    for e in example_updates.iter() {
        updates.add_event(e.clone()).unwrap();
    }

    let seralized_bytes: Vec<u8> = bincode::serialize(&updates).unwrap();
    let deseralized_update: GraphUpdate = bincode::deserialize(&seralized_bytes).unwrap();

    assert_eq!(0, deseralized_update.len().unwrap());
    assert_eq!(true, deseralized_update.is_empty().unwrap());
}

#[test]
fn serialize_json() {
    let example_updates = vec![
        UpdateEvent::AddNode {
            node_name: "parent".into(),
            node_type: "corpus".into(),
        },
        UpdateEvent::AddNode {
            node_name: "child".into(),
            node_type: "corpus".into(),
        },
        UpdateEvent::AddEdge {
            source_node: "child".into(),
            target_node: "parent".into(),
            layer: "annis".into(),
            component_type: "PartOf".into(),
            component_name: "".into(),
        },
    ];

    let mut updates = GraphUpdate::new();
    for e in example_updates.iter() {
        updates.add_event(e.clone()).unwrap();
    }

    let seralized_string = serde_json::to_string_pretty(&updates).unwrap();
    assert_snapshot!(seralized_string);
}

#[test]
fn serialize_deserialize_json() {
    let example_updates = vec![
        UpdateEvent::AddNode {
            node_name: "parent".into(),
            node_type: "corpus".into(),
        },
        UpdateEvent::AddNode {
            node_name: "child".into(),
            node_type: "corpus".into(),
        },
        UpdateEvent::AddEdge {
            source_node: "child".into(),
            target_node: "parent".into(),
            layer: "annis".into(),
            component_type: "PartOf".into(),
            component_name: "".into(),
        },
    ];

    let mut updates = GraphUpdate::new();
    for e in example_updates.iter() {
        updates.add_event(e.clone()).unwrap();
    }

    let seralized_string = serde_json::to_string_pretty(&updates).unwrap();
    let deseralized_update: GraphUpdate = serde_json::from_str(&seralized_string).unwrap();

    assert_eq!(3, deseralized_update.len().unwrap());
    let deseralized_events: Vec<UpdateEvent> = deseralized_update
        .iter()
        .unwrap()
        .map(|e| e.unwrap().1)
        .collect();
    assert_eq!(example_updates, deseralized_events);
}
