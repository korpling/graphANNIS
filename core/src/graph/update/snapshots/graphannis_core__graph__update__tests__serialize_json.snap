---
source: core/src/graph/update/tests.rs
expression: seralized_string
---
{
  "1": {
    "AddNode": {
      "node_name": "parent",
      "node_type": "corpus"
    }
  },
  "2": {
    "AddNode": {
      "node_name": "child",
      "node_type": "corpus"
    }
  },
  "3": {
    "AddEdge": {
      "source_node": "child",
      "target_node": "parent",
      "layer": "annis",
      "component_type": "PartOf",
      "component_name": ""
    }
  }
}
